t/e2e/whisper/whisper_mailbox_test.go

```diff
@@ -296,6 +300,195 @@ func (s *WhisperMailboxSuite) TestRequestMessagesInGroupChat() {
+	}
+
+	// wait for mailserver to archive all the envelopes
+	s.waitForEnvelopeEvents(envelopeArchivedWatcher, sentEnvelopesHashes, whisper.EventMailServerEnvelopeArchived)
+
+	// first page
+	// send request
+	requestID := requestMessages("")
+	// wait for mail server response
+	resp := s.waitForMailServerResponse(mailServerResponseWatcher, requestID)
+	s.NotEmpty(resp.LastEnvelopeHash)
+	s.NotEmpty(resp.Cursor)
+	// wait for last envelope sent by the mailserver to be available for filters
+	s.waitForEnvelopeEvents(envelopeAvailableWatcher, []string{resp.LastEnvelopeHash.String()}, whisper.EventEnvelopeAvailable)
+	// get messages
+	firstPageHashes := getMessages()
+	s.Equal(3, len(firstPageHashes))
+
+	// second page
+	// send request
+	requestID = requestMessages(fmt.Sprintf("%x", resp.Cursor))
+	// wait for mail server response
+	resp = s.waitForMailServerResponse(mailServerResponseWatcher, requestID)
+	s.NotEmpty(resp.LastEnvelopeHash)
+	// all messages have been sent, no more pages available
+	s.Empty(resp.Cursor)
+	// wait for last envelope sent by the mailserver to be available for filters
+	s.waitForEnvelopeEvents(envelopeAvailableWatcher, []string{resp.LastEnvelopeHash.String()}, whisper.EventEnvelopeAvailable)
+	// get messages
+	secondPageHashes := getMessages()
+	s.Equal(2, len(secondPageHashes))
+
+	allReceivedHashes := append(firstPageHashes, secondPageHashes...)
+	s.Equal(envelopesCount, len(allReceivedHashes))
+
+	// check that all the envelopes have been received
+	sort.Strings(sentEnvelopesHashes)
+	sort.Strings(allReceivedHashes)
+	s.Equal(sentEnvelopesHashes, allReceivedHashes)
+}
+
+func (s *WhisperMailboxSuite) waitForEnvelopeEvents(events chan whisper.EnvelopeEvent, hashes []string, event whisper.EventType) {
+	check := make(map[string]struct{})
+	for _, hash := range hashes {
+		check[hash] = struct{}{}
+	}
+
+	timeout := time.NewTimer(time.Second * 5)
+	for {
+		select {
+		case e := <-events:
+			if e.Event == event {
+				delete(check, e.Hash.String())
+				if len(check) == 0 {
+					timeout.Stop()
+					return
+				}
+			}
+		case <-timeout.C:
+			s.FailNow("timed out while waiting for event on envelopes", "event: %s", event)
+		}
+	}
+}
+
+func (s *WhisperMailboxSuite) waitForMailServerResponse(events chan whisper.EnvelopeEvent, requestID common.Hash) *whisper.MailServerResponse {
+	timeout := time.NewTimer(time.Second * 5)
+	for {
+		select {
+		case event := <-events:
+			if event.Event == whisper.EventMailServerRequestCompleted && event.Hash == requestID {
+				timeout.Stop()
+				resp, ok := event.Data.(*whisper.MailServerResponse)
+				if !ok {
+					s.FailNow("mailserver response error", "expected whisper.MailServerResponse, got: %+v", resp)
+				}
+
+				return resp
+			}
+		case <-timeout.C:
+			s.FailNow("timed out while waiting for mailserver response")
+		}
+	}
+}
+
+func (s *WhisperMailboxSuite) addPeerAndWait(node, other *node.StatusNode) {
```

Comment:
In `t/helpers/peers.go` there are some helper functions to wait for an added peer.